#### 编译时异常和运行时异常

> 异常就是代表程序出现的问题

![](E:\the-notes-of-itheima\Image\异常.png)

+ 编译时异常：除了RuntimeException和他的其他子类，其他都是编译时异常。编译阶段需要进行处理，作用在于提醒程序员。
+ 运行时异常：RuntimeException本身和所有子类，都是运行时异常。编译阶段不报错，是程序运行时出现的。

#### 作用

+ 作用一：异常是用来查询bug的关键参考信息
+ 作用二：异常可以作为方法内部的一种特殊返回值，以便通知调用者底层的执行情况

```java
public void setAge ( int age){
            if (age < 18 || age > 40) {
                throw new RuntimeException();
            } else {
                this.age = age;
            }
        }
```

#### 处理方式

1. JVM默认的处理方式

   > 异常信息输出在控制台，剩余代码停止运行

2. 自己处理（捕获异常）

格式：

```java
try {
    可能出现异常的代码;
}catch(想要捕获的异常类名 变量名){
    异常的处理代码;
}
```

代码示例：

```java
int[] arr={1,2,3,4,5};
        try{
            //可能出现异常的代码
            System.out.println(arr[10]);
        }catch (ArrayIndexOutOfBoundsException e){
            //如果出现了异常，应该怎么去处理异常
            System.out.println("索引越界了");
        }
```

好处：当代码出现异常时，可以让程序继续往下执行

注意事项：

+ 当不同异常处理的方案不同时，一个catch只能捕获一个异常。
+ 当多个异常处理的方案相同时，一个catch可以写多个异常，中间用一个 | 分隔。
+ catch语句可以多个一起用。
+ 如果要捕获多个异常，这些异常中如果存在父子关系的话，那么父类一定要卸载下面，因为catch执行时是从上往下依次执行的。
+ 当try...catch没有捕获异常时，会按照虚拟机默认处理方案进行。
+ try中写了多行语句，如果上面的语句发生异常并被捕获，那下面的代码就不会继续执行而是跳转到对应的catch中，如果没有对应的catch，就会进行虚拟机默认的处理方案。

Throwable的成员方法

| 方法名称                      | 说明                            |
| ----------------------------- | ------------------------------- |
| public String getMessage()    | 返回此throwable的详细消息字符串 |
| public String toString()      | 返回此处可抛出的简短描述        |
| public void printStackTrace() | 把异常的错误信息输出在控制台    |

+ printStackTrace仅仅是把异常的错误信息输出在控制台，不会中止代码的进行，区别于虚拟机的默认处理方案

3. 抛出处理

+ throws：写在方法定义处，表示说明一个异常，告诉调用者使用本方法可能会有哪些异常

  ```java
   public void 方法() throws 异常类名1，异常类名2...{
              ...
          }
  ```

  + 编译时异常必须写，运行时异常可以不写

+ throw：写在方法内，结束方法，手动抛出异常对象并交给调用者，方法下面的代码不再执行

  ```java
         public void 方法(){
                  throw new 异常类名();
          }
  ```

  

代码示例：

```java
public static int getMax(int[] arr) {
        if (arr == null) {
            throw new NullPointerException();
            //以下代码不再进行
        }
        int max = arr[0];
        for (int i = 0; i < arr.length; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }
```

#### 自定义异常

+ 意义：就是为了让控制台的报错信息更加的见名知意

1. 定义异常类

2. 写继承关系

   > 运行时异常继承RuntimeException
   >
   > 编译时异常继承Exception

3. 空参构造

4. 带参构造

```java
public class NameFormatException extends RuntimeException{
    public NameFormatException() {
    }
    public NameFormatException(String message) {
        super(message);
    }
}
```

